src/mongocxx/options/index.hpp

```diff
@@ -0,0 +1,361 @@
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+class collection;
+
+namespace options {
+
+///
+/// Class representing the optional arguments to a MongoDB create index operation.
+///
+/// @see http://docs.mongodb.org/manual/reference/method/db.collection.createIndex/#options
+///
+class MONGOCXX_API index {
+   public:
+    ///
+    /// Base class representing the optional storage engine options for indexes.
+    ///
+    class base_storage_options {
+       private:
+        friend collection;
+        MONGOCXX_PRIVATE virtual const int type() const = 0;
+    };
+
+    ///
+    /// Class representing the optional WiredTiger storage engine options for indexes.
+    ///
+    class wiredtiger_storage_options : public base_storage_options {
+       public:
+        ///
+        /// Set the WiredTiger configuration string.
+        ///
+        /// @param config_str
+        ///   The WiredTiger configuration string.
+        ///
+        void config_str(std::string config_str);
+
+        ///
+        /// The current config_str setting.
+        ///
+        /// @return The current config_str.
+        ///
+        const stdx::optional<std::string>& config_str() const;
+
+       private:
+        friend collection;
+        MONGOCXX_PRIVATE const int type() const override;
+        stdx::optional<std::string> _config_str;
+    };
+
+    index();
+
+    ///
+    /// Whether or not to build the index in the background so that building
+    /// the index does not block other database activities.
+    ///
+    /// @param background
+    ///   Whether or not to build the index in the background.
+    ///
+    /// @see https://docs.mongodb.org/manual/tutorial/build-indexes-in-the-background/
+    ///
+    void background(bool background);
+
+    ///
+    /// The current background setting.
+    ///
+    /// @return The current background.
+    ///
+    const stdx::optional<bool>& background() const;
+
+    ///
+    /// Whether or not to create a unique index so that the collection will not
+    /// accept insertion of documents where the index key or keys match an
+    /// existing value in the index.
+    ///
+    /// @param unique
+    ///   Whether or not to create a unique index.
+    ///
+    void unique(bool unique);
+
+    ///
+    /// The current unique setting.
+    ///
+    /// @return The current unique.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-unique/
+    ///
+    const stdx::optional<bool>& unique() const;
+
+    ///
+    /// The name of the index.
+    ///
+    /// @param name
+    ///   The name of the index.
+    ///
+    void name(std::string name);
+
+    ///
+    /// The current name setting.
+    ///
+    /// @return The current name.
+    ///
+    const stdx::optional<std::string>& name() const;
+
+    ///
+    /// Whether or not to create a sparse index. Sparse indexes only reference
+    /// documents with the indexed fields.
+    ///
+    /// @param sparse
+    ///   Whether or not to create a sparse index.
```

Comment:
Probably good to link the sparse index page, too: https://docs.mongodb.org/manual/core/index-sparse/